[cryptotest] Test RSA sign using wycheproof vectors

nasahlpa opened the pull request:
Test the sign operation using the wycheproof vectors.

nasahlpa committed:
[cryptotest] Add `rsa_sign` handler

Extend the cryptotest firmware to be able to perform a RSA sign operation.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[cryptotest] Add wycheproof RSA sign vectors

As CL only supports public exponents of e = 65537, reuse some of the wycheproof
test vectors for checking the sign operation.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[cryptotest] Add RSA sign host harness

Check whether the sign operation was correct by doing a verify operation
afterwards. This is needed as other existing sign test vectors (e.g., from
NIST do not use e = 65537 but CL only supports that public exponent.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[cryptotest] Add RSA Wycheproof Sign testing

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa commented:
All sign tests pass:
```
351:Finished test test_rsa: Ok(())
```

andrea-caforio reviewed (approved):
From October 21, nice. :-)

siemen11 commented on sw/device/tests/crypto/cryptotest/BUILD:
```diff
@@ -233,6 +233,41 @@ RSA_TESTVECTOR_TARGETS = [
+        (
+            2048,
+            [
+                "sha256",
```
Are there any sha3 hashes in there?

nasahlpa replied:
There are no native SHA3 test vectors with the public exponent we use available. However, as the hashing is actually agnostic for the verify after sign approach, we simply can re-use some of the SHA-2 test vectors, switch to SHA-3 and check if the verify passes at the end with the same hashing mode.

siemen11 replied:
Gotcha, this is done in the functest as well. I was wondering if there were more vectors, thanks for checking!

siemen11 reviewed (approved):
Thank you Pascal!

johannheyszl reviewed (approved):
thanks @nasahlpa

nasahlpa committed:
[cryptotest] Add RSA sign SHA-3 test vectors

Although there are no test vectors available that use SHA-3, reuse
some SHA-2 test vectors and manually override the mode to SHA-3.
This works as the verify-after-sign testing approach enable us to
do this.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa commented:
Merging as all sign tests pass:
```
350:Finished test test_rsa: Ok(())
```
